watch/watch.go

```diff
@@ -0,0 +1,360 @@
+// Copyright 2017 The OPA Authors.  All rights reserved.
+// Use of this source code is governed by an Apache2
+// license that can be found in the LICENSE file.
+
+package watch
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/open-policy-agent/opa/ast"
+	"github.com/open-policy-agent/opa/rego"
+	"github.com/open-policy-agent/opa/storage"
+)
+
+// Watcher allows for watches to be registed on queries. When a dependency of a watched
+// query changes, a notification will be send back to the user.
+type Watcher struct {
```

Comment:
This is not a huge deal. I'd recommend spiking the interface and then writing example tests that show usage for the simple case as well as migration. Let's see what they look like. I think the separation might lead to more idiomatic code.